### PR TITLE
add tooltip to unit_attack dialogs

### PR DESCRIPTION
**Pull request by newfrenchy83**
in tooltip, the special and weapon_abilities are listed by stats (damage, attacks and chance to hit) and owner of each is specified.
replace https://github.com/wesnoth/wesnoth/pull/6399 for readibility purpose.

### CODE REVIEW

File: src/units/abilities.cpp
Comment (stevecotton) [inline src/units/abilities.cpp:902]: What do weapon_abilities and checking_name have to do with adding a string? This seems to be part of the flow logic of the caller instead.

File: src/units/abilities.cpp
Comment (stevecotton) [inline src/units/abilities.cpp:915]: I can't follow the logic of this function - it seems that `weapon_abilities` is a temporary string that's used for assembling parts of other strings called `self_abilities` etc, and if that's what it is then I don't see why `weapon_abilities` is in scope for the whole function. I also can't see why it's an input parameter to the weapon_specials_*_self functions, as it seems to be an output from those.

However, this line uses it purely as an output, and IIUC the value set here would be discarded, were it not for the first weapon_specials_impl_self call taking it as an input, and then possibly concatenating other values to it before copying it to self_abilities. However, that happens in a conditional block that only happens if self_ is non-null (which might always be true, but then why is there a conditional block)?

**Review by stevecotton — CHANGES_REQUESTED** [review]

File: src/units/abilities.cpp
Comment (newfrenchy83) [reply src/units/abilities.cpp:961]: why, checking_tags is constant inside function, reference taht not pertinent because value not changed.

Comment (stevecotton) [reply src/units/abilities.cpp:961]: Because passing by value makes an unnecessary copy of it.

File: src/units/abilities.cpp
Comment (newfrenchy83) [reply src/units/abilities.cpp:963]: excuse me for this condition, this is because i forget what this in attack dialog, self_ and other_ are ever true, but i had implement that by reflexe.

File: src/units/abilities.cpp
Comment (newfrenchy83) [reply src/units/abilities.cpp:902]: i must rename that clear_temp_string instead.

File: src/units/abilities.cpp
Comment (newfrenchy83) [reply src/units/abilities.cpp:915]: weapon_abilities used mike reference,in add_name, i cannot put other stinrg in input or i put the code itself. I will try ths option.

File: src/units/abilities.cpp
Comment (newfrenchy83) [reply src/units/abilities.cpp:961]: Ok

File: src/units/abilities.cpp
Comment (newfrenchy83) [reply src/units/abilities.cpp:963]: but i will let this condition, else i fear what building not pass checking

File: src/units/abilities.cpp
Comment (stevecotton) [reply src/units/abilities.cpp:902]: Irrespective of the names, what do these two variables have to do with this function; why are they involved in it? They're not inputs, and they don't really feel like outputs either - they both get changed, but their final value doesn't depend on any of the inputs to the function.

Comment (stevecotton) [reply src/units/abilities.cpp]: Reviewing ea7bcc0c452, I still don't understand why checking_name is involved in this function.

Comment (newfrenchy83) [reply src/units/abilities.cpp]: This is for improbable, but not impossible, cases where weapons from different sources (an ability carried by self, and another by teacher for example), would have the same name, if in the window, the name should only be mentioned only once, in the tooltip, it seems relevant to point out that the stats are influenced by two weapons possessed and taught, the example of a charge type ability can also illustrate this when two units fight with the same capacity of this type (active both in attack and defense), the player sees the attack stats of his unit influenced by his own ability but also that of his opponent. To allow this, checking_name must be cleared after each block (self, teacher or opponent) is written.

Comment (stevecotton) [reply src/units/abilities.cpp]: I think I understand, but feel it could be refactored to be simpler. If I understand the current implementation correctly:
* When `add_name_list` returns, `checking_name` is always empty. Although the `clear()` is in a conditional block, that condition will always be true except for the times that `checking_name` is already empty when `add_name_list` is entered.
* In `weapon_specials_value` the only times that `checking_name` could be non-empty on entry to a `weapon_specials_impl_*` function are the ones for the `"Self: "` and `"Opponent: "` blocks.
* There are more places that it could be non-empty in `weapon_specials`.
* In `weapon_specials_value` there's only one time that the value of `checking_name` returned from a `weapon_specials_impl_*` function is used, for the `"Opponent: "` block.

Comment (newfrenchy83) [reply src/units/abilities.cpp]: for the first point, conditionnal block not ever true, if weapon_specials_impl_* return a empty temp_string, and preffix not added to non-existent list.
. Correct "Self: " has weapon specials and abilities like weapon owned by unit in list, and opponent has all weapon special or abilities like weapon applied used by opponent(owned or teached) and applied to unit

. Correct weapon_specials apllied to windows and origin to each weapon\abilities) non specifed, a name must be addedone time only.

. Yes because opponent is the only case or two weapon_specials_impl_* are used.

Comment (stevecotton) [reply src/units/abilities.cpp]: I'm wondering if we can omit special handling for this edge-case. For the `"Self: "` case, I think it only happens if the unit has an ability called `x` and a weapon special also called `x` - in which case, showing `x, x` will look odd but possibly be easier for the player to understand than combining them into `x` when two badly-named things are active.

For the `"Other: "` string, please would you give an example of an ability that should be returned by `weapon_specials_impl_adj`? I'm wondering if those will already be in "Enemy's Teachers: ".

Comment (newfrenchy83) [reply src/units/abilities.cpp]: using "Self: " for abilities only, ok. 

For "Opponent: " a leadership charge teached to opponent but appied also to attacker must appear in in tooltip, and the "opponent: " string is here for said what damage *2by abilities used by opponent, same if teached.

**Comment by newfrenchy83** [comment]
@stevecotton i removed only_active argument to attack_type::weapon_specials()

**Comment by newfrenchy83** [comment]
@stevecotton PR rebased

**Comment by newfrenchy83** [comment]
@stevecotton it is long time what PR not reviewed.

**Comment by newfrenchy83** [comment]
@Vultraz it is a long time what this PR not reviewed, could you give an advice, see what this PR is a UI work.

**Comment by Pentarctagon** [comment]
Can you post a screenshot of what it looks like with this PR's change applied?

**Comment by newfrenchy83** [comment]
> Can you post a screenshot of what it looks like with this PR's change applied?

teached: (cantor encoded in Elvish Captain and leadership)
tooltip_teached

teached by an enemy(diversion ability)
tooltip_teachedenemy


used by opponent(formation ability)
tooltip_usedopponent

**Comment by newfrenchy83** [comment]
@Pentarctagon  @Vultraz , i have a question, with abilities/specials showed in tooltip, show the weapon teached by enemy or used by opponent in attack prediction windows is useful, or must i  show in tooltip only for not give heavy the window itself, and show in window weapon owned or teached by allies?

**Comment by Pentarctagon** [comment]
I'd say to keep it just in the tooltip for now. Given how long this PR has already been open already, let's not add more to it.

**Comment by newfrenchy83** [comment]
> I'd say to keep it just in the tooltip for now. Given how long this PR has already been open already, let's not add more to it.

except what 'diversion' and formation both in window in this PR, must becommit a remove of that or don't commit anymore?

**Comment by newfrenchy83** [comment]
@Pentarctagon i reformulate my question: 'diversion' and 'formation' name must appear in window(without mention of origin) or not? And if not, must commit that in this PR or i another PR after merging in master?

**Comment by Pentarctagon** [comment]
I think it'd be best to see what vultraz says first.

Just as far as wording goes though, for correctness:
* "teached" should be changed to "taught" in the above screenshots.
* "enemy" is misspelled as "ennemy".
* "others aspect" should be "other aspects".

**Comment by newfrenchy83** [comment]
@Pentarctagon excuse my lack of patience but i would post remording in PR, and i let what weapon owned or taught by allies in window.

**Review by Vultraz — APPROVED** [review]

**Comment by stevecotton** [comment]
I apologise for how long it's taken me to review this, and thank @Vultraz for reviewing the UI.

My plan is to merge this evening, unless anyone has more to say. If I have any more ideas about possible refactoring, I'll create a new PR on top of this instead of blocking this PR.

**Pull request merged**
